### PR TITLE
[Feature/Identity] Introduce negated permission parsing and integrate guice

### PR DESCRIPTION
**Pull request by scrawfor99**
### Description
This PR will add negated permission parsing and incorporate the glob matching logic for resolving permission implications. 

### Issues Resolved
[List any issues this PR will resolve]

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: sandbox/modules/identity/src/main/java/org/opensearch/identity/authz/OpenSearchPermission.java [hunk 1]
```diff
@@ -10,27 +10,169 @@
...
+    public boolean permissionTypesMatch(OpenSearchPermission permission) {
+
+        if (this.permissionType.equalsIgnoreCase(permission.permissionType)) {
+            return true;
         }
+        return false;
+    }
+
+    /**
+     * Resolves complex resource patterns with '-' signs to the corresponding resource patterns
```
Comment (peternied) [inline sandbox/modules/identity/src/main/java/org/opensearch/identity/authz/OpenSearchPermission.java:93]: Nice to see an implementation, but just want to voice reservations about adding negative permissions.

IMO I'd rather be able to decompose an incoming request like `GET /index-6*,-index-666` so we don't check for permission of `index-666`, but avoid letting declaration the permissions in that way.

Comment (scrawfor99) [reply sandbox/modules/identity/src/main/java/org/opensearch/identity/authz/OpenSearchPermission.java:93]: Yeah, I know in the past you had mentioned being hesitant on this. I just wanted to put this here and show it working so we can check it off the todos. It is something that had been discussed with permission granting and I would rather have it ready to go should we decide it is worthwhile then not have it. I understand your thinking though and think that I would be happy with either.

**Review by cwperks — COMMENTED** [review]
Hey @scrawfor99, thank you for the PR. To help in reviewing the PR, can you add tests and more of a description to show what you are thinking?

To add to what Peter mentioned, I think that negative permissions would cause complexity and its also not something I see supported in the security plugin currently (Look for instances of `index_permissions` in the security repo - https://github.com/search?q=repo%3Aopensearch-project%2Fsecurity%20index_permissions&type=code)

When granting permissions like "Grant permission to perform foo and bar, but not baz" it can also be stated as "Grant permission to perform foo and bar". With wildcards it can be a little trickier, but it should be possible to represent any combination of permissions to concrete indices using positive permissions grants.

**Comment by scrawfor99** [comment]
> Hey @scrawfor99, thank you for the PR. To help in reviewing the PR, can you add tests and more of a description to show what you are thinking?
> 
> To add to what Peter mentioned, I think that negative permissions would cause complexity and its also not something I see supported in the security plugin currently (Look for instances of `index_permissions` in the security repo - https://github.com/search?q=repo%3Aopensearch-project%2Fsecurity%20index_permissions&type=code)
> 
> When granting permissions like "Grant permission to perform foo and bar, but not baz" it can also be stated as "Grant permission to perform foo and bar". With wildcards it can be a little trickier, but it should be possible to represent any combination of permissions to concrete indices using positive permissions grants.

Hi Craig, thank you for your comments. This is the next thing in-line after the permissions API. There will be tests in the future but for now, there are not because I have not been actively working on this branch. The inclusion of the negated indices is just an experimental change to see how it would work. It was something discussed in the design conversation on 1/27 so I figured I would quickly include it in the draft in case we did want it.

**Pull request closed without merging**